Hybrid Sharded Data Parallel

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89915

Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- @awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- @liangluofb For ideation, feedback, and initial implementation and experimentation

rohan-varma committed:
HSDP

[ghstack-poisoned]

rohan-varma committed:
Update on "HSDP"




[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/api.py:
```diff
@@ -46,12 +46,22 @@ class ShardingStrategy(Enum):
       :class:`DistributedDataParallel` API. For gradients, this strategy
       synchronizes them (via all-reduce) after the backward computation. The
       unsharded optimizer states are updated locally per rank.
+    - ``HYBRID_SHARD``: Apply ``FULL_SHARD`` within a node, and data parallelism across
```
```suggestion
    - ``HYBRID_SHARD``: Apply ``FULL_SHARD`` within a node and replicate parameters across
```

`FULL_SHARD` is still implementing data parallelism; only it is _sharded_ data parallelism.

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/api.py:
```diff
@@ -46,12 +46,22 @@ class ShardingStrategy(Enum):
       unsharded optimizer states are updated locally per rank.
+    - ``HYBRID_SHARD``: Apply ``FULL_SHARD`` within a node, and data parallelism across
+        nodes. This results in reduced communication volume as expensive all-gathers and
+        reduce-scatters are only done within a node, and can be more performant for medium
```
```suggestion
        reduce-scatters are only done within a node, which can be more performant for medium-
```

(`medium-sized` with a hyphen)

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/api.py:
```diff
@@ -46,12 +46,22 @@ class ShardingStrategy(Enum):
+        nodes. This results in reduced communication volume as expensive all-gathers and
+        reduce-scatters are only done within a node, and can be more performant for medium
+        sized models.
+    - ``HYBRID_SHARD_ZERO2``: Apply ``SHARD_GRAD_OP`` within a node, and data parallelism across
```
Same comment as above:
```suggestion
    - ``HYBRID_SHARD_ZERO2``: Apply ``SHARD_GRAD_OP`` within a node and replicate parameters across
```

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/api.py:
```diff
@@ -46,12 +46,22 @@ class ShardingStrategy(Enum):
     NO_SHARD = auto()
-    # HYBRID_SHARD = auto()
+    HYBRID_SHARD = auto()
+    HYBRID_SHARD_ZERO2 = auto()
```
As discussed offline, let us make this private. We should revisit naming if we want to make this public.

awgu replied:
[Blocking] Let us change this before landing

awgu commented on torch/distributed/fsdp/api.py:
```diff
@@ -46,12 +46,22 @@ class ShardingStrategy(Enum):
+        reduce-scatters are only done within a node, and can be more performant for medium
+        sized models.
+    - ``HYBRID_SHARD_ZERO2``: Apply ``SHARD_GRAD_OP`` within a node, and data parallelism across
+        nodes. This results in reduced communication volume as expensive all-gathers and
```
nit: I think we should omit this sentence "This results in ..." since it is the same as for `HYBRID_SHARD`. Instead, we can say something like
```
This is like ``HYBRID_SHARD``, except this may provide even higher 
throughput since the unsharded parameters are not freed after the
forward pass, saving the all-gathers in the pre-backward.
```

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/api.py:
```diff
@@ -46,12 +46,22 @@ class ShardingStrategy(Enum):
+        nodes. This results in reduced communication volume as expensive all-gathers and
+        reduce-scatters are only done within a node, and can be more performant for medium
+        sized models. Compared to ``HYBRID_SHARD``, users may wish to try this for even smaller
+        models that may not have to free full parameters after the forward pass.
```
nit: Let us refer to the parameters are "sharded" or "unsharded" and slowly retire "full".
```suggestion
        models that may not have to free the unsharded parameters after the forward pass.
```

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -609,9 +663,14 @@ def _should_free_in_backward(
+    # We always free if we are syncing gradients (i.e. not in no_sync) and parameters
+    # are sharded.
+    free_unsharded = state._sync_gradients and handle.uses_sharded_strategy
+    # For NO_SHARD we don't need to free full parameters, for Zero-2 strategies, we skip
```
We should have a discussion about if we want to refer to `reshard_after_forward = False` strategies as ZeRO-2. I am personally not a big fan because ZeRO-2 is not accurate, and it also requires users to learn about ZeRO-2/3 differences to understand our APIs, which will then also lead them to more confusion since they do not align.

awgu replied:
[Not blocking] However, we should definitely discuss sometime

awgu replied:
[Not blocking] However, I recommend us think about how to refer to the `reshard_after_forward = False` strategies. The debt will only build with more PRs on top of this.

rohan-varma replied:
agreed, we should consider this in follow up discussions

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -548,6 +593,15 @@ def _post_backward_hook(
                     padded_unsharded_grad,
                     new_sharded_grad,
                 )
+                if handle._config.sharding_strategy in {
```
nit: No point incurring the cost of constructing a `set` here if there are only 2 elements. `tuple` is fine.

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -29,6 +30,40 @@
+            )
+
+        # Ensure inter and intra-node process groups are the same
+        # TODO (rohan-varma) unclear whether these should be asserts or Exceptions
```
Are we intending to resolve this TODO before landing?

My opinion is to make these exceptions and make our own FSDP code solid to ensure an incorrect configuration cannot happen.

awgu replied:
[Not blocking] However, I do not see why not address this in this PR before landing.

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+@no_type_check
+def _init_intra_node_process_group() -> dist.ProcessGroup:
+    """
+    Returns a process group across the current machine.
```
nit: Shall we be consistent about language and call it a "node", not a "machine"? (unless I am misunderstanding what a "machine" is here)

```suggestion
    Returns a process group across the current node.
```

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+def _init_intra_node_process_group() -> dist.ProcessGroup:
+    """
+    Returns a process group across the current machine.
+    For example, given each row is a distinct machine:
```
Maybe we should give this example in terms of 8 GPUs / node, which is typical of hardware right now.

awgu replied:
[Not blocking] It is still worth thinking about though. If we do not change it in this PR, I doubt we will ever change it.

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
     return state
 
+@no_type_check
+def _init_process_group_state_for_hybrid_shard(state, process_group) -> _FSDPState:
```
Is it worth type annotating `process_group` here? Even adding `Any` may be valuable to show that we need to validate the typing in this function.

awgu replied:
[Easy] Let us type annotate this actually

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+            )
+        else:
+            state = _init_process_group_state_for_hybrid_shard(state, process_group)
+            assert state.process_group is not None, "Expected to populate state.process_group for hybrid shard"
```
Do we really want these asserts? This seems like we can just read `_init_process_group_state_for_hybrid_shard` to make sure these are set.

awgu replied:
[Not blocking] But please think about if we need/want these `assert`s

rohan-varma replied:
I think these are useful, it's to make sure that `_init_process_group_state_for_hybrid_shard` sets all of these attributes.

In the future, `_init_process_group_state_for_hybrid_shard` could possibly be refactored or modified to set these attributes in helper functions, then reading through that would become a bit more cumbersome.

awgu replied:
I think the asserts should be at the end of `_init_process_group_state_for_hybrid_shard` then, representing a post-condition. This is just my personal design preference and probably does not matter in this case. However, in general, you would have to enforce the post-condition upon each call to `_init_process_group_state_for_hybrid_shard`.

(This is also how I approach post-conditions and invariants in general -- they should be coupled to the method/function itself, not their usages.)

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+    if sharding_strategy in HYBRID_SHARDING_STRATEGIES:
+        if process_group is None and policy is None:
+            # Raise an error here, since this is manual wrapping with no process group
+            # passed in, there is no way to ensure all wrapped FSDP units use the same
```
nit: Shall we stop using "units" and converge to "instances" (as in a Python object instance)?

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+            raise ValueError(
+                f"process_group should be None or dist.ProcessGroup, but got {type(process_group)}"
+            )
+        if process_group is None:
+            state.process_group = _get_default_group()
+        else:
+            state.process_group = process_group
```
```suggestion
        state.process_group = process_group or _get_default_group()
```

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+    return state
+
+@no_type_check
+def _validate_hybrid_shard_pg_type(process_group: Any) -> bool:
```
nit: Since this returns a `bool`:
```suggestion
def _is_valid_hybrid_shard_pg_type(process_group: Any) -> bool:
```

If the function is called `_validate...`, then I would expect it to return `None` and only perform asserts/checks.

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
     return state
 
+@no_type_check
+def _init_process_group_state_for_hybrid_shard(state, process_group) -> _FSDPState:
```
Type annotate `state`?

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+        state._inter_node_pg = inter_node_group
+    else:
+        # Check type and assign state.process_group and state._inter_node_pg.
+        valid = _validate_hybrid_shard_pg_type(process_group)
+        if valid:
```
Following other comment below:
```suggestion
        if _is_valid_hybrid_shard_pg_type(process_group):
```

(I like this kind of naming because then it reads like English.)

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+    global_process_group: dist.ProcessGroup
+) -> dist.ProcessGroup:
+    """
+    Returns an inter-node process group where each contained rank has
```
Same comments as for intra-node method

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,140 @@
+    world_size = dist.get_world_size(global_process_group)
+    # Assuming fully homogeneous setup
+    num_devices = torch.cuda.device_count()
+    num_distinct_nodes = world_size // num_devices
```
nit: Like my comment on the original PR,
```suggestion
    num_nodes = world_size // num_devices
```
(Does "distinct" mean something different than if it was not present?)

awgu replied:
[Easy] I recommend changing before landing

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -29,6 +30,40 @@
+}
+
+@no_type_check
+def _validate_hybrid_shard_setup(state: _FSDPState, fsdp_module: nn.Module):
```
Can we skip this check altogether for composable code path?

In that case, should we change the first argument to `fsdp_root` or something?

rohan-varma replied:
curious why we need to skip this check for the composable path?

awgu replied:
The composable path does not support manual wrapping, so there should be no way for the process groups to be inconsistent. In other words, there should be no need to validate the setup. Given this, I recommend writing this method with function argument names that reflect the fact that it is only meant for the non-composable code path.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -364,6 +368,12 @@ def __init__(
                 # FSDP module directly
                 submodule._fsdp_use_orig_params = use_orig_params
 
+        # Initializes self.process_group, along with rank and world size. This will
```
nit: Personally, I do not like explaining what a function/method call does inline like this since this creates redundancy, which can go stale if only one place is updated. The developer should read the docstring for `_init_process_group_state`.

awgu replied:
[Easy] I recommend changing before landing

rohan-varma replied:
I mostly want to emphasize the part a couple lines later that mentions this is done before auto wrapping, and the logic for why, which I think is vaulable.

awgu replied:
Sounds good.

awgu reviewed (commented):
I made an initial pass and left a lot of nitpicks. I will read the test code in a follow-up pass, possibly after you respond to some of the comments.

awgu commented:
Just as a heads up, it looks like test failures are real:
```
======================================================================
ERROR [0.002s]: test_wrap_wrap_method_WrapMethod_WRAP_API (__main__.TestAutoWrap)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 247, in instantiated_test
    test(self, **param_kwargs)
  File "/var/lib/jenkins/workspace/test/distributed/fsdp/test_wrap.py", line 319, in test_wrap
    layer = wrap(nn.Linear(5, 5))
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/wrap.py", line 295, in wrap
    return _wrap(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/wrap.py", line 313, in _wrap
    return wrapper_cls(module, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 376, in __init__
    _init_process_group_state(self, process_group, sharding_strategy, auto_wrap_policy)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/fsdp/_init_utils.py", line 110, in _init_process_group_state
    raise ValueError(
ValueError: process_group should be None or dist.ProcessGroup, but got <class 'torch.testing._internal.common_fsdp.DummyProcessGroup'>
```

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -189,8 +190,11 @@ class FullyShardedDataParallel(nn.Module):
+            This is the process group used for collective communications and
+            the one over which the model is sharded. For hybrid sharding strategies such as
+            ``ShardingStrategy.HYBRID_SHARD`` or ``ShardingStrategy.HYBRID_SHARD_ZERO2``, users can
+            pass in a tuple of process groups representing the sharding and data parallel groups.
```
```suggestion
            pass in a tuple of process groups representing the groups to shard and replicate across, respectively.
```

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/api.py:
```diff
@@ -46,12 +46,21 @@ class ShardingStrategy(Enum):
+        nodes. This results in reduced communication volume as expensive all-gathers and
+        reduce-scatters are only done within a node, which can be more performant for medium
+        -sized models.
+    - ``_HYBRID_SHARD_ZERO2``: Apply ``SHARD_GRAD_OP`` within a node, and replicate parameters across
```
I guess we should omit this from docstring for now

awgu replied:
Yes, I think that would be good to be safe.

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,245 @@
+@contextlib.contextmanager
+def patch_allreduce(new_allreduce):
+    """
+    Patches dist.all_reduce with a new all_reduce and
```
nit (no need to change): Some interesting inconsistencies between `all_reduce` and `allreduce` 🤔

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,245 @@
+class TestFSDPHybridShard(FSDPTest):
+    @property
+    def world_size(self):
+        return max(torch.cuda.device_count(), 2)
```
Curious: What is the purpose of this `max()` if we skip if less than 2 GPUs?

rohan-varma replied:
This was mostly to make CI happy, when there are no cuda devices, this would return 0, which would cause issues in test setup before the tests are skipped.

awgu replied:
Do you know why we do not need to do this for any of the other FSDP test files? (E.g. is it due to some process group initialization specific to this file?)

rohan-varma replied:
This was mostly to fix a CI issue, when there are no GPUs but this code still runs.

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,245 @@
+        """
+        Tests basic functionality of HYBRID_SHARD and _HYBRID_SHARD_ZERO2:
+            1. Inter and intra-node process groups are correctly setup
+            2. Process groups are the same across FSDP wrapped units
```
```suggestion
            2. Process groups are the same across FSDP wrapped instances
```

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,245 @@
+        for sharding_strategy in [
+            ShardingStrategy.HYBRID_SHARD, ShardingStrategy._HYBRID_SHARD_ZERO2
+        ]:
+            with self.subTest(sharding_strategy=sharding_strategy):
```
Eventually, we should refactor use `self.run_subtests()` which wraps over this `self.subTest` and includes the final `dist.barrier()`.

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,147 @@
+def _init_intra_node_process_group() -> dist.ProcessGroup:
+    """
+    Returns a process group across the current node.
+    For example, given each column is a distinct node:
```
nit: To save lines/verticality, can we transpose this array? :)

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -75,13 +88,147 @@
+    """
+    Returns an inter-node process group where each contained rank has
+    the same local rank. For example, given each column is a distinct node:
+    0 8
```
Same: Can we transpose?

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -609,9 +666,14 @@ def _should_free_in_backward(
+    # We always free if we are syncing gradients (i.e. not in no_sync) and parameters
+    # are sharded.
+    free_unsharded = state._sync_gradients and handle.uses_sharded_strategy
+    # For NO_SHARD we don't need to free full parameters, for Zero-2 strategies, we skip
```
nit: Let us stylize following the paper?
```suggestion
    # For NO_SHARD we don't need to free full parameters, for ZeRO-2 strategies, we skip
```

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -609,9 +666,14 @@ def _should_free_in_backward(
+    # are sharded.
+    free_unsharded = state._sync_gradients and handle.uses_sharded_strategy
+    # For NO_SHARD we don't need to free full parameters, for Zero-2 strategies, we skip
+    # freeing in backward. Note that this also applies for _HYBRID_SHARD_ZERO2.
```
nit: This note seems pretty clear? Otherwise, which are the ZeRO-2 _strategies_ (plural)?
```suggestion
    # freeing in backward. Note that this also applies for _HYBRID_SHARD_ZERO2.
```

rohan-varma replied:
Shard grad op and _HYBRID_SHARD_ZERO2 right?

awgu replied:
What I mean is that you already said in the comment "for Zero-2 strategies" (plural), and (at least to me) `SHARD_GRAD_OP` is the obvious first one. Then, since you said plural "strategies", there must be a second one, and the second one is probably `_HYBRID_SHARD_ZERO2` since that has `ZERO2` in the name.

This is just a nitpick saying in my opinion we do not need to clarify with the "Note that this also applies for _HYBRID_SHARD_ZERO2" since the preceding sentence is already clear.

awgu reviewed (commented):
I have a final pass of nits, and we can land after that.

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -307,11 +312,28 @@ def init(
+                "sharding_strategy" in fsdp_kwargs
+                and fsdp_kwargs["sharding_strategy"] in {
+                    ShardingStrategy.HYBRID_SHARD,
+                    ShardingStrategy.HYBRID_SHARD_ZERO2
```
Missed one (which is causing the test failures):
```suggestion
                    ShardingStrategy._HYBRID_SHARD_ZERO2
```

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

awgu reviewed (approved):
Looks great to me! Awesome work, and thanks for fixing all of the nits. I am very excited to see the experiment results and downstream impact!

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -0,0 +1,244 @@
+# Owner(s): ["oncall: distributed"]
+
+import contextlib
+import functools
```
```suggestion
```

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma committed:
Update on "Hybrid Sharded Data Parallel"


Adds 2 new hybrid sharding strategy to FSDP:
1. HYBRID_SHARD: applies zero-3 style sharding within a node, and data parallel across
2. HYBRID_SHARD_ZERO2: applies zero-2 style sharding within a node, and data parallel across

These are useful for medium sized models and aim to decrease communication volume, tests and benchmarks will be run to understand which workloads are optimal under which sharding strategy.

Hybrid sharding in general works by sharding the model using a process group within a single node, and creating intra-node process groups for replication / data parallelism. The user either needs to pass in a tuple of these process groups, or None, and we generate the process groups appropriately.

** Acknowledgements **
- awgu 's excellent prototype: https://github.com/awgu/pytorch/commit/5ad3a16d486484c9ab4445126e50655eb19d62ca
- liangluofb For ideation, feedback, and initial implementation and experimentation

[ghstack-poisoned]

rohan-varma commented:
CI failures are related to autocast and are unrelated to this PR.

rohan-varma commented:
@pytorchbot merge -f "CI failures unrelated"

jeffdaily commented:
This PR broke ROCm CI periodic jobs, where the distributed tests get run.

```
======================================================================
ERROR [4.530s]: test_fsdp_hybrid_shard_basic_setup (__main__.TestFSDPHybridShard)
Tests basic functionality of HYBRID_SHARD and _HYBRID_SHARD_ZERO2:
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 533, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 759, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 804, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 657, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 535, in wrapper
    fn()
  File "/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 166, in wrapper
    return func(*args, **kwargs)
  File "/var/lib/jenkins/pytorch/test/distributed/fsdp/test_fsdp_hybrid_shard.py", line 226, in test_fsdp_hybrid_shard_basic_setup
    with (
AttributeError: __enter__
```

jeffdaily commented:
Not sure how this PR only broke ROCm CI, unless this test is getting skipped on other platforms?

The test is using the context managers from earlier in the file, so not sure why `__enter__` isn't an attribute?

https://github.com/pytorch/pytorch/pull/89915/files#diff-19a0c73c8366fbd896a68ad90a2f4b3515f0fa6dc0a61528a2780e418b3f5e92R44

awgu commented:
> Not sure how this PR only broke ROCm CI, unless this test is getting skipped on other platforms?
> 
> The test is using the context managers from earlier in the file, so not sure why `__enter__` isn't an attribute?
> 
> https://github.com/pytorch/pytorch/pull/89915/files#diff-19a0c73c8366fbd896a68ad90a2f4b3515f0fa6dc0a61528a2780e418b3f5e92R44

This could be because of a Python versioning issue. Parenthesized context managers may have only been added in Python 3.10: https://www.blog.pythonlibrary.org/2021/09/08/python-3-10-parenthesized-context-managers/

In other words, the following syntax was not permitted until 3.10:
```
with (
    patch_allreduce(patched_allreduce),
    patch_reduce_scatter(patched_reduce_scatter),
):
```
This could be why the `__enter__` is not being detected properly.

jeffdaily commented:
@awgu I just figured out the same.  PR to fix.  https://github.com/pytorch/pytorch/pull/90580
